include/dr/mp/algorithms/count.hpp

```diff
@@ -0,0 +1,136 @@
+// SPDX-FileCopyrightText: Intel Corporation
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+namespace dr::mp::__detail {
+
+inline auto add_counts(rng::forward_range auto &&r) {
+  rng::range_difference_t<decltype(r)> zero{};
+
+  return std::accumulate(rng::begin(r), rng::end(r), zero);
+}
+
+inline auto count_if_local(rng::forward_range auto &&r, auto &&pred) {
+  if (mp::use_sycl()) {
+    dr::drlog.debug("  with DPL\n");
+#ifdef SYCL_LANGUAGE_VERSION
+    return std::count_if(mp::dpl_policy(),
+                         dr::__detail::direct_iterator(rng::begin(r)),
+                         dr::__detail::direct_iterator(rng::end(r)), pred);
+#else
+    assert(false);
+#endif
+  } else {
+    dr::drlog.debug("  with CPU\n");
+    return std::count_if(std::execution::par_unseq,
+                         dr::__detail::direct_iterator(rng::begin(r)),
+                         dr::__detail::direct_iterator(rng::end(r)), pred);
+  }
+}
+
+template <dr::distributed_range DR>
+auto count_if(std::size_t root, bool root_provided, DR &&dr, auto &&pred) {
+  using count_type = rng::range_difference_t<decltype(dr)>;
+  auto comm = mp::default_comm();
+
+  if (rng::empty(dr)) {
+    return count_type{};
+  }
+
+  if (aligned(dr)) {
```

Comment:
One range is always aligned with itself. Remove this condition and "else" branch.